hax/treemakers/posrec.py

```diff
@@ -0,0 +1,163 @@
+from scipy.stats import binom_test
+import json
+from keras.models import model_from_json, Sequential
+from keras.layers import Dense, Dropout, Activation, Reshape, Flatten
+from keras import regularizers
+
+class PositionReconstruction(TreeMaker):
+    """Stores position-reconstruction-related variables.
+    
+    Provides:
+       - s1_pattern_fit: s1 pattern fit computed with corrected
+                         position and areas
+       - s1_pattern_fit_hits: s1 pattern fit computed with corrected
+                         position and hits
+
+    """
+    __version__ = '0.12'
+    extra_branches = ['peaks.area_per_channel[260]',
+                      'peaks.hits_per_channel[260]',
+                      'peaks.n_saturated_per_channel',
+                      'interactions.x','interactions.y','interactions.z']
+
+    def __init__(self):
+        
+        hax.minitrees.TreeMaker.__init__(self)
+        
+        # We need to pull some stuff from the pax config
+        self.pax_config = load_configuration("XENON1T")
+        self.tpc_channels = list(range(0,   247+1))
+        self.confused_s1_channels = []
+        self.statistic = self.pax_config \
+                         ['BuildInteractions.BasicInteractionProperties'] \
+                         ['s1_pattern_statistic']
+        qes = np.array(self.pax_config['DEFAULT']['quantum_efficiencies'])
+
+        self.pattern_fitter = PatternFitter(
+            filename=utils.data_file_name(
+                self.pax_config['WaveformSimulator']['s1_patterns_file']),
+            zoom_factor = self.pax_config['WaveformSimulator'].get(
+                's1_patterns_zoom_factor', 1),
+            adjust_to_qe=qes[self.tpc_channels],
+            default_errors=(self.pax_config['DEFAULT']['relative_qe_error'] +
+                            self.pax_config['DEFAULT']['relative_gain_error'])
+        )
+
+        # For s1_aft variable
+        # Don't yell at me for hardcoding the filename into hax because it was
+        # also hardcoded in pax. Just kicking the can.
+        #aftmap_filename = utils.data_file_name('s1_aft_xyz_XENON1T_06Mar2017.json')
+        aftmap_filename = utils.data_file_name('s1_aft_xyz_XENON1T_20170808.json')
+        self.aft_map = InterpolatingMap(aftmap_filename)
+        self.low_pe_threshold=10
+        #hax.minitrees.Treemaker.__init__(self)
+        # load trained NN models
+        nn_model_json = utils.data_file_name('tensorflow_nn_pos_XENON1T_20171211.json')
+        json_file_nn = open(nn_model_json, 'r')
+        loaded_model_json = json_file_nn.read()
+        json_file_nn.close()
+        loaded_nn_model = model_from_json(loaded_model_json)
+        weights_file = utils.data_file_name('tensorflow_nn_pos_weights_XENON1T_20171211.h5')
+        loaded_nn_model.load_weights(weights_file)
+        self.nn_tensorflow = loaded_nn_model
+        self.list_bad_pmts = [1, 2, 12, 26, 34, 62, 65, 79, 86, 88, 102, 118, 130, 134, 135, 139, 148, 150, 152, 162, 178,
```

Comment:
can this be taken from the PMT gain list above?

Reply:
Why needed? Bad pmts are blinded in pax (gain zero) so have no signal in the hit pattern.

Reply:
@feigaodm says: defines the number of layers in NN.

Hardcoded for SR0 and SR1 now in https://github.com/XENON1T/hax/pull/184. @weiyuehuan working on including this list with the NN models to replace this.

Reply:
fixed in https://github.com/XENON1T/hax/commit/c994b089d11ed6ab3210b91d32028677dc745796 and https://github.com/XENON1T/hax/pull/174/commits/c1849fd8db8a5d8d7597b2a1abed1f605d6ac9ce